File: src/contest/ccc/CCC_2014_S5.java
package contest.ccc;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.StringTokenizer;

public class CCC_2014_S5 {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;

	public static void main (String[] args) throws IOException {
		int n = nextInt();
		int[][] points = new int[n + 1][];
		points[0] = new int[] {0, 0};
		int counter = 0;
		int[][] pairs = new int[n * (n + 1) / 2][];
		for (int x = 1; x < n + 1; x++) {
			if (x != 0)
				points[x] = new int[] {nextInt(), nextInt()};
			for (int y = x - 1; y >= 0; y--) {
				int x1 = points[x][0] - points[y][0];
				int y1 = points[x][1] - points[y][1];
				pairs[counter] = new int[] {y, x, x1 * x1 + y1 * y1};
				counter++;
			}
		}

		// for(int x = 0; x < n+1; x++){
		// for(int y = x+1; y < n+1; y++){
		// int x1 = points[x][0]-points[y][0];
		// int y1 = points[x][1]-points[y][1];
		// pairs[counter] = new int[]{x,y,x1*x1+y1*y1};
		// counter++;
		// }
		// }
		Arrays.sort(pairs, new Comparator<int[]>() {
			@Override
			public int compare (int[] arg0, int[] arg1) {
				return arg0[2] - arg1[2];
			}
		});
		/*
		 * for(int[] x:pairs){ System.out.println(x[0] + " " + x[1] + " " +
		 * x[2]); }
		 */
		int[] bestDist = new int[n + 1];
		int[] prevbestNum = new int[n + 1];
		int[] bestNum = new int[n + 1];
		for (int x = 0; x < pairs.length; x++) {
			int[] next = pairs[x];
			if (next[2] > bestDist[next[0]]) {
				bestDist[next[0]] = next[2];
				prevbestNum[next[0]] = bestNum[next[0]];
			}
			if (next[2] > bestDist[next[1]]) {
				bestDist[next[1]] = next[2];
				prevbestNum[next[1]] = bestNum[next[1]];
			}
			// adding necessary previous row memorization in order to calc curr
			// row
			if (next[0] == 0) {

				bestNum[next[0]] = Math.max(bestNum[next[0]], prevbestNum[next[1]]);
			} else {

				bestNum[next[0]] = Math.max(bestNum[next[0]], prevbestNum[next[1]] + 1);
				bestNum[next[1]] = Math.max(bestNum[next[1]], prevbestNum[next[0]] + 1);
			}
			/*
			 * for(int y = 0; y < n+1; y++){ System.out.print(bestNum[y] + ":" +
			 * bestDist[y] + " "); } System.out.println();
			 */
		}
		System.out.println(bestNum[0] + 1);
	}

	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());

		return st.nextToken();
	}

	static int nextInt () throws IOException {
		return Integer.parseInt(next());
	}

	static String nextLine () throws IOException {
		return br.readLine().trim();
	}
}

File: src/contest/ccc/CCC_2014_S5.cc
#include <bits/stdc++.h>

using namespace std;

int n;

struct point {
  int x;
  int y;
  point(int x, int y) {
    this->x = x;
    this->y = y;
  }
};

struct line {
  int i1;
  int i2;
  int dist;
  line(int i1, int i2, int dist) {
    this->i1 = i1;
    this->i2 = i2;
    this->dist = dist;
  }
};

vector<point> points;
int bestDist[2001];
int prevbestNum[2001];
int bestNum[2001];
vector<line> lines;

bool compare(line p1, line p2) {
  return p2.dist > p1.dist;
}

int main() {
  scanf("%d", &n);
  points.push_back(point(0, 0));
  int count = 0;
  for (int x = 1; x <= n; x++) {
    if (x != 0) {
      int a, b;
      scanf("%d%d", &a, &b);
      points.push_back(point(a, b));
    }
    for (int y = x - 1; y >= 0; y--) {
      int x1 = points[x].x - points[y].x;
      int y1 = points[x].y - points[y].y;
      lines.push_back(line(y, x, x1 * x1 + y1 * y1));
      count++;
    }
  }
  sort(lines.begin(), lines.begin() + (n) * (n + 1) / 2, compare);
  for (int x = 0; x < (n) * (n + 1) / 2; x++) {
    line next = lines[x];
    if (next.dist > bestDist[next.i1]) {
      bestDist[next.i1] = next.dist;
      prevbestNum[next.i1] = bestNum[next.i1];
    }
    if (next.dist > bestDist[next.i2]) {
      bestDist[next.i2] = next.dist;
      prevbestNum[next.i2] = bestNum[next.i2];
    }
    // adding necessary previous row memorization in order to calc curr row
    if (next.i1 == 0) {
      bestNum[next.i1] = max(bestNum[next.i1], prevbestNum[next.i2]);
    } else {
      bestNum[next.i1] = max(bestNum[next.i1], prevbestNum[next.i2] + 1);
      bestNum[next.i2] = max(bestNum[next.i2], prevbestNum[next.i1] + 1);
    }
  }
  printf("%d", bestNum[0] + 1);
  return 0;
}
